### PR TITLE
[ONNX] Fix symbolic values and numpy implementation

### LINKED ISSUE #135700: [ONNX][high-pri] ONNX symbolic values are unhashable; numpy() implementation is incomplete
cc @gramalingam

### PR DESCRIPTION
**Pull request by justinchuby**
1. Remove `__eq__` to make `SymbolicTensor` hashable and test for that
2. Update the `__array__` method so that it works for tensor on GPU

cc @gramalingam @xadupre 

Fixes https://github.com/pytorch/pytorch/issues/135700

### CODE REVIEW

**Review by titaiwangms — APPROVED** [review]

File: torch/onnx/_internal/exporter/_core.py [hunk 1]
```diff
@@ -99,17 +99,18 @@ def __init__(self, tensor: torch.Tensor, name: str | None = None):
 
     def __array__(self, dtype: Any = None) -> np.ndarray:
         # numpy() calls __array__ in ir.Tensor
+        self.raw: torch.Tensor
         if self.dtype == ir.DataType.BFLOAT16:
-            return self.raw.view(torch.uint16).__array__(dtype)
+            return self.raw.view(torch.uint16).numpy(force=True).__array__(dtype)
```
Comment (xadupre) [inline torch/onnx/_internal/exporter/_core.py:104]: Maybe for a future change, is there a way to keep torch.Tensor?

Comment (justinchuby) [reply torch/onnx/_internal/exporter/_core.py:104]: The tensor can be accessed via .raw, or have its content copied via dlpack. The array magic will always be numpy array. Does that sound good to you?

Comment (xadupre) [reply torch/onnx/_internal/exporter/_core.py:104]: Yes, I should have asked my second question. What is used when the model is serialized and for constant folding?

Comment (justinchuby) [reply torch/onnx/_internal/exporter/_core.py:104]: For serialization, tobytes is called. For constant folding, numpy is used because we use the reference implementation

Comment (xadupre) [reply torch/onnx/_internal/exporter/_core.py:104]: I see. ``tensor = self.raw.detach().cpu().contiguous()`` should probably be written with contiguous first ``tensor = self.raw.contiguous().detach().cpu()``. It is not necessarily a cheap. For the reference evaluator, maybe we can update it to support torch Tensor (through the array API) so that the cost if cheaper as well. That will wait for the time being.

Comment (justinchuby) [reply torch/onnx/_internal/exporter/_core.py:104]: According to https://pytorch.org/docs/stable/generated/torch.Tensor.numpy.html  this is the same as t.detach().cpu().resolve_conj().resolve_neg().numpy() so contiguous is not called?

Comment (xadupre) [reply torch/onnx/_internal/exporter/_core.py:104]: I'm saying it should happen on cuda not cpu.

Comment (justinchuby) [reply torch/onnx/_internal/exporter/_core.py:104]: resolve conj and resolve neg will materialize some flags I think. I am speculating torch didn’t want to change the original tensor.

Comment (justinchuby) [reply torch/onnx/_internal/exporter/_core.py:104]: But judging from the nature of the tensors we are handling (weights) they will most likely be no-ops.

Comment (gramalingam) [reply torch/onnx/_internal/exporter/_core.py:104]: Not sure I follow the above discussion ... I don't see any call to detach(), is that not required? Anyway: with regards to perf implications, we should evaluate and decide and that is a separate issue for the future.

Comment (justinchuby) [reply torch/onnx/_internal/exporter/_core.py:104]: The force=True option will call everything above including detach
